src/core/sio.cc

```diff
@@ -809,31 +787,83 @@ void PCSX::SIO::GetMcdBlockInfo(int mcd, int block, McdBlock *Info) {
     trim(str);
     trim(sstr);
 
-    ptr = data + block * 8192 + 0x60;  // icon palette data
-
-    for (i = 0; i < 16; i++) {
-        clut[i] = *((unsigned short *)ptr);
-        ptr += 2;
-    }
+    // Read CLUT
+    ptr = data + block * MCD_BLOCK_SIZE + 0x60;
+    std::memcpy(clut, ptr, 16 * sizeof(uint16_t));
 
-    for (i = 0; i < Info->IconCount; i++) {
+    // Icons can have 1 to 3 frames of animation
+    for (int i = 0; i < Info->IconCount; i++) {
         uint16_t *icon = &Info->Icon[i * 16 * 16];
+        ptr = data + block * MCD_BLOCK_SIZE + 128 + 128 * i;  // icon data
 
-        ptr = data + block * 8192 + 128 + 128 * i;  // icon data
-
+        // Fetch each pixel, store it in the icon array in ABBBBBGGGGGRRRRR with the alpha bit set to 1
         for (x = 0; x < 16 * 16; x++) {
-            icon[x++] = clut[*ptr & 0xf];
-            icon[x] = clut[*ptr >> 4];
+            const uint8_t entry = (uint8_t)*ptr;
+            icon[x++] = clut[entry & 0xf] | (1 << 15);
+            icon[x] = clut[entry >> 4] | (1 << 15);
             ptr++;
         }
     }
 
-    ptr = data + block * 128;
+    
+    // Parse directory frame info
+    const auto directoryFrame = (uint8_t *)data + block * MCD_SECT_SIZE;
+    Info->Flags = directoryFrame[0];
+    std::strncpy(Info->ID, (const char*) &directoryFrame[0xa], 12);
+    std::strncpy(Info->Name, (const char*) &directoryFrame[0x16], 16);
+
+    const uint32_t fileSize =
+        directoryFrame[0x4] | (directoryFrame[0x5] << 8) | (directoryFrame[0x6] << 16) | (directoryFrame[0x7] << 24);
+    Info->Filesize = fileSize;
+
+    // Check if the block is marked as free in the directory frame and adjust the name/filename if so
+    if (Info->Flags == 0xa0) {
+        std::strcpy(Info->Title, "Free Block");
+        std::strcpy(Info->Name, "Empty File");
+        Info->Filesize = 0;
+    }
+}
 
-    Info->Flags = *ptr;
+char *PCSX::SIO::GetMcdData(int mcd) {
```

Comment:
Maybe `mcd` should be of an enum type.

Reply:
Does it really make sense to use something like
```cpp
enum class MemoryCards {
    Card1, Card2, ...
};
```
When that would more or less be the exact same thing as using an integer but with more steps? An argument could also be made about `int` allowing us to support an arbitrary amount of memory cards eventually, but I don't think that's really useful.

Reply:
The "more steps" is sort of precisely the point. Well, there's two reasons. One is to improve readability of calls into the API, without having to look at the prototype or the comments, in order to avoid calls in an obtuse way like

```c
doSomethingWithMcd(0, 12, 52, 28);
```

and have something a bit better like

```c
doSomethingWithMcd(MCD::Slot::S1, MCD::Command::XYZ, 52, 28);
```

It's okay to try and reduce the number of letters to type tho, I agree on this.

The second reason is to avoid basic errors like putting the arguments in the wrong order, by forcing a stronger C++ type for each argument as much as possible.